### ISSUE TITLE
not_unsafe_ptr_arg_deref with function in an extern block 

### ISSUE THREAD

**Issue by indefini** [rust-lang/rust-clippy#12134]
### Summary

The function is in an extern block but the lint says it is not marked as unsafe.

### Lint Name

not_unsafe_ptr_arg_deref

### Reproducer

I tried this code:

```rust
extern "C" {
    fn test(p : *const i32);
}

pub fn test2(t: *const i32) {
    // this doesn't compile because test is unsafe
    //test(t);
    // clippy complains about this because test is not marked unsafe
    unsafe { test(t) };
}

```
playground : https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4149c9acb12f5c6ba6dfc6cb112f9627

I saw this happen:

```
error: this public function might dereference a raw pointer but is not marked `unsafe`
```

### Version

```text
rustc 1.75.0 (82e1608df 2023-12-21)
binary: rustc
commit-hash: 82e1608dfa6e0b5569232559e3d385fea5a93112
commit-date: 2023-12-21
host: x86_64-unknown-linux-gnu
release: 1.75.0
LLVM version: 17.0.6
```

**Comment by y21** [comment]
The lint wants the `test2` function to be marked as unsafe, because there isn't a way for this `test2` function to check that the pointer is valid before using it, so it has to trust the caller to pass a valid pointer. When this is the case, the function should be marked unsafe to require an unsafe block at call site.

With the function written as is, a user can write `test2(1234 as *const i32);` and cause undefined behavior in safe code.

**Comment by indefini** [comment]
@y21 Thank you.
I thought that the lint wanted test (not test2) to be marked as unsafe.
I am closing this.
